Resolve #412 - Negative damage must be ignored

viniciuslrangel opened the pull request:
When an AttackableUnit receives damage (TakeDamage) less than 0, raises the damage to 0

Refs #412

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/LeagueSandbox/GameServer?pullRequest=748) <br/>All committers have signed the CLA.

viniciuslrangel commented:
Ops, sorry, I forgot to send the packet to show the 0 text.

MythicManiac commented on GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -128,6 +128,10 @@ public virtual bool IsInDistress()
             DamageText damageText)
         {
+            if (damage < 0f) {
+                Logger.Error("Damage must be non-negative");
```
The log entry is a bit excessive don't you think? There are ways to deal negative damage in vanilla League of Legends as well, that shouldn't be an issue.

viniciuslrangel replied:
I guess that this would be a fault from the spell/item/etc code.

MythicManiac replied:
Also now that I think about it, this probably should be implemented by setting the damage to 0 rather than returning from the function. There might be cases where we need to track how many instances of damage taking events have passed, so 0 damage should follow through the same logic as 100 damage.

viniciuslrangel replied:
done

MythicManiac commented:
Otherwise looks good, but please refer to our [committing guidelines](https://github.com/LeagueSandbox/GameServer/wiki/4.07.-Committing) on how to format your commits.

moonshadow565 commented on GameServerLib/GameObjects/AttackableUnits/AttackableUnit.cs:
```diff
@@ -167,6 +167,10 @@ public virtual bool IsInDistress()
+                damage = 0f;
+            }
             //Damage dealing. (based on leagueoflegends' wikia)
             damage = defense >= 0 ? 100 / (100 + defense) * damage : (2 - 100 / (100 - defense)) * damage;
```
Might want to move the check after finnal damage calculation

viniciuslrangel committed:
Resolve #412 -- Raises negative damage to 0

To avoid negative expressions (and healing instead of damage), raises the damage to make final result equals to 0

Refs #412

danil179 commented:
If damage<0 it will remain smaller than 0 so checking at the start is sort of optimization.

You can prove it by cases, assuming damage<0 at the start:
defense>0:
Then of course 100/(100+defense)>0 so (100/(100+defense))*damage<0
defense<0:
100-defense>100 => 0<100/(100-defense)<1=> 2>2 - 100/(100-defense)>1
Hence (2 - 100/(100-defense))*damage<0

Also easy to prove that damage=0 remain 0.

Therefore, I merge this.

danil179 reviewed (approved):
Approved.

We should track what sends negative damage at the future (although it could be league system), currently it is sort of adhoc solution.
